File: src/perception/scripts/plot_perception_kpi.py
#!/usr/bin/env python3
"""
plot_perception_kpi.py

Generate 3 publication-ready KPI plots for the Zooba perception system final report.
Reads traffic_light_kpi.csv and sign_detection_kpi.csv from ~/zooba_kpi/ and saves
PNG plots to ~/zooba_kpi/plots/.
"""

import os
import csv
import sys
import numpy as np
import matplotlib.pyplot as plt

def read_csv(filepath):
    data = {}
    if not os.path.exists(filepath):
        print(f"Warning: file not found {filepath}")
        return data
    with open(filepath, 'r') as f:
        reader = csv.reader(f)
        try:
            headers = next(reader)
        except StopIteration:
            return data
            
        for h in headers:
            data[h.strip()] = []
        for row in reader:
            if len(row) != len(headers):
                continue
            for i, val in enumerate(row):
                data[headers[i].strip()].append(val.strip())
    return data

def moving_average(a, n=50):
    if len(a) == 0:
        return np.array([])
    ret = np.cumsum(a, dtype=float)
    ret[n:] = ret[n:] - ret[:-n]
    return ret[n - 1:] / n

def plot_perception_kpi(kpi_dir):
    tl_file = os.path.join(kpi_dir, 'traffic_light_kpi.csv')
    sign_file = os.path.join(kpi_dir, 'sign_detection_kpi.csv')
    
    tl_data = read_csv(tl_file)
    sg_data = read_csv(sign_file)
    
    if not tl_data and not sg_data:
        print("No CSV files found or empty in:", kpi_dir)
        return

    # Colors
    c_tl_det = '#348abd' # Steel blue
    c_tl_trk = '#1cba8b' # Emerald green
    c_sign = '#db604f'   # Coral red
    
    # Process TL
    if tl_data and 'timestamp' in tl_data:
        tl_time = np.array([float(x) for x in tl_data['timestamp']])
        if len(tl_time) > 0:
            tl_time = tl_time - tl_time[0]
        
        tl_latency = np.array([float(x) for x in tl_data['latency_ms']])
        tl_mode = np.array(tl_data['mode'])
        tl_det_lat = np.array([float(x) for x in tl_data['detection_latency_ms']])
        tl_trk_lat = np.array([float(x) for x in tl_data['tracking_latency_ms']])
        
        det_latencies = tl_det_lat[tl_mode == 'DETECTION']
        trk_latencies = tl_trk_lat[tl_mode == 'TRACKING']
        
        tl_rate = np.array([float(x) for x in tl_data['detection_rate_pct']])
        tl_ratio = np.array([float(x) for x in tl_data['best_color_ratio']])
    else:
        det_latencies, trk_latencies = np.array([]), np.array([])
        tl_time, tl_rate, tl_ratio = np.array([]), np.array([]), np.array([])

    # Process SG
    if sg_data and 'timestamp' in sg_data:
        sg_time = np.array([float(x) for x in sg_data['timestamp']])
        if len(sg_time) > 0:
            sg_time = sg_time - sg_time[0]
            
        sign_latencies = np.array([float(x) for x in sg_data['latency_ms']])
        sg_rate = np.array([float(x) for x in sg_data['detection_rate_pct']])
        sg_conf = np.array([float(x) for x in sg_data['vote_confidence']])
    else:
        sign_latencies = np.array([])
        sg_time, sg_rate, sg_conf = np.array([]), np.array([]), np.array([])

    out_dir = os.path.join(kpi_dir, 'plots')
    os.makedirs(out_dir, exist_ok=True)
    
    # Helper to compute stats safely
    def get_stats(arr):
        if len(arr) == 0: return 0.0, 0.0, 0.0, 0.0
        return np.mean(arr), np.std(arr), np.percentile(arr, 95), np.max(arr)
        
    det_mean, det_std, det_p95, det_max = get_stats(det_latencies)
    trk_mean, trk_std, trk_p95, trk_max = get_stats(trk_latencies)
    sign_mean, sign_std, sign_p95, sign_max = get_stats(sign_latencies)

    # ══════════════════════════════════════════════════════════════
    # FIGURE 1 - Latency Bar Chart
    # ══════════════════════════════════════════════════════════════
    plt.figure(figsize=(10, 6))
    
    labels = ['TL Detection', 'TL Tracking', 'Sign Detection']
    means = [det_mean, trk_mean, sign_mean]
    stds = [det_std, trk_std, sign_std]
    colors = [c_tl_det, c_tl_trk, c_sign]
    
    x_pos = np.arange(len(labels))
    plt.bar(x_pos, means, yerr=stds, color=colors, alpha=0.8, capsize=10, width=0.6)
    plt.axhline(y=50, color='#800000', linestyle='--', linewidth=1.5, label='50 ms real-time')
    
    plt.ylabel('Latency (ms)', fontweight='bold', fontsize=12)
    plt.title('Perception Pipeline Latency', fontweight='bold', fontsize=14)
    plt.xticks(x_pos, labels, fontweight='bold', fontsize=11)
    plt.grid(axis='y', linestyle='--', alpha=0.7)
    
    stats_str = (
        f"TL Detection:\n  Mean={det_mean:.1f}  Std={det_std:.1f}  P95={det_p95:.1f}  Max={det_max:.1f}\n\n"
        f"TL Tracking:\n  Mean={trk_mean:.1f}  Std={trk_std:.1f}  P95={trk_p95:.1f}  Max={trk_max:.1f}\n\n"
        f"Sign Detection:\n  Mean={sign_mean:.1f}  Std={sign_std:.1f}  P95={sign_p95:.1f}  Max={sign_max:.1f}"
    )
    # Position the text box appropriately so it doesn't overlap too much
    plt.text(1.3, max([m+s for m,s in zip(means, stds)] + [50]) * 0.4, stats_str, 
             fontsize=10, family='monospace', 
             bbox=dict(facecolor='white', alpha=0.9, edgecolor='gray'))
    
    plt.legend()
    plt.tight_layout()
    plt.savefig(os.path.join(out_dir, 'latency_bar_chart.png'), dpi=300)
    plt.close()

    # ══════════════════════════════════════════════════════════════
    # FIGURE 2 - Detection Rate
    # ══════════════════════════════════════════════════════════════
    plt.figure(figsize=(10, 5))
    if len(tl_time) > 0:
        plt.plot(tl_time, tl_rate, label='Traffic Light', color=c_tl_det, linewidth=2)
        plt.plot(tl_time[-1], tl_rate[-1], 'o', color=c_tl_det, markersize=8)
        plt.annotate(f"{tl_rate[-1]:.1f}%", (tl_time[-1], tl_rate[-1]), 
                     xytext=(-10, 10), textcoords='offset points', color=c_tl_det, fontweight='bold')
        
    if len(sg_time) > 0:
        plt.plot(sg_time, sg_rate, label='Sign Detection', color=c_sign, linewidth=2)
        plt.plot(sg_time[-1], sg_rate[-1], 'o', color=c_sign, markersize=8)
        plt.annotate(f"{sg_rate[-1]:.1f}%", (sg_time[-1], sg_rate[-1]), 
                     xytext=(-10, -15), textcoords='offset points', color=c_sign, fontweight='bold')
        
    plt.axhline(y=80, color='gray', linestyle=':', linewidth=1.5, label='Target 80%')
    plt.ylim(0, 105)
    plt.xlabel('Time (s)', fontweight='bold', fontsize=12)
    plt.ylabel('Detection Rate (%)', fontweight='bold', fontsize=12)
    plt.title('Cumulative Detection Rate', fontweight='bold', fontsize=14)
    plt.legend(loc='lower right')
    plt.grid(linestyle='--', alpha=0.7)
    plt.tight_layout()
    plt.savefig(os.path.join(out_dir, 'detection_rate.png'), dpi=300)
    plt.close()

    # ══════════════════════════════════════════════════════════════
    # FIGURE 3 - Confidence
    # ══════════════════════════════════════════════════════════════
    fig, ax1 = plt.subplots(figsize=(10, 5))
    
    if len(tl_time) > 0:
        ax1.plot(tl_time, tl_ratio, color=c_tl_det, linewidth=1, alpha=0.5, label='TL Colour Ratio')
        if len(tl_ratio) >= 50:
            smooth_tl = moving_average(tl_ratio, 50)
            ax1.plot(tl_time[49:], smooth_tl, color=c_tl_det, linewidth=2, label='TL (smoothed)')
            
    ax1.set_xlabel('Time (s)', fontweight='bold', fontsize=12)
    ax1.set_ylabel('TL Colour Ratio', fontweight='bold', fontsize=12, color=c_tl_det)
    ax1.tick_params(axis='y', labelcolor=c_tl_det)
    ax1.set_ylim(0, 1.05)
    
    ax2 = ax1.twinx()
    if len(sg_time) > 0:
        ax2.plot(sg_time, sg_conf, color=c_sign, linewidth=1, alpha=0.5, label='Sign Confidence')
        if len(sg_conf) >= 50:
            smooth_sg = moving_average(sg_conf, 50)
            ax2.plot(sg_time[49:], smooth_sg, color=c_sign, linewidth=2, label='Sign (smoothed)')
            
    ax2.set_ylabel('Sign Vote Confidence', fontweight='bold', fontsize=12, color=c_sign)
    ax2.tick_params(axis='y', labelcolor=c_sign)
    ax2.set_ylim(0, 1.05)
    
    plt.title('Detection Confidence', fontweight='bold', fontsize=14)
    
    lines_1, labels_1 = ax1.get_legend_handles_labels()
    lines_2, labels_2 = ax2.get_legend_handles_labels()
    ax1.legend(lines_1 + lines_2, labels_1 + labels_2, loc='upper center', bbox_to_anchor=(0.5, -0.15), ncol=4)
    
    ax1.grid(linestyle='--', alpha=0.7)
    plt.tight_layout()
    plt.savefig(os.path.join(out_dir, 'confidence.png'), dpi=300)
    plt.close()

    print(f"\n✓ Plots saved to {out_dir}")
    print("  - latency_bar_chart.png")
    print("  - detection_rate.png")
    print("  - confidence.png")

if __name__ == '__main__':
    kpi_dir = os.path.expanduser('~/zooba_workspace/zooba_kpi') # Adjust default dir
    if len(sys.argv) > 1:
        kpi_dir = sys.argv[1]
    plot_perception_kpi(kpi_dir)

File: src/perception/scripts/plot_perception_kpi.m
function plot_perception_kpi(kpi_dir)
% PLOT_PERCEPTION_KPI  Generate 3 publication-ready KPI plots for the
%                      Zooba perception system final report.
%
%   plot_perception_kpi()          — uses ~/zooba_kpi/ as default
%   plot_perception_kpi(kpi_dir)   — custom path to KPI CSV folder
%
%   Reads:
%       <kpi_dir>/traffic_light_kpi.csv
%       <kpi_dir>/sign_detection_kpi.csv
%
%   Produces 3 figures (saved as PNG):
%       1. Latency bar chart   (TL Detection, TL Tracking, Sign Detection)
%       2. Detection rate      (TL and Sign over time)
%       3. Confidence          (TL colour ratio and Sign vote confidence)
%
%   Authors: Zooba Autonomous Vehicle Team

    if nargin < 1
        kpi_dir = fullfile(getenv('HOME'), 'zooba_kpi');
    end

    tl_file   = fullfile(kpi_dir, 'traffic_light_kpi.csv');
    sign_file = fullfile(kpi_dir, 'sign_detection_kpi.csv');

    % ── Colour palette ────────────────────────────────────────────
    C.tl_det   = [0.204 0.541 0.741];   % Steel blue
    C.tl_trk   = [0.110 0.725 0.545];   % Emerald green
    C.sign     = [0.859 0.380 0.310];   % Coral red
    C.grid     = [0.85  0.85  0.85 ];   % Light grey grid
    C.bg       = [0.98  0.98  0.98 ];   % Near-white background
    C.err      = [0.3   0.3   0.3  ];   % Dark grey for error bars

    % ── Read CSVs ─────────────────────────────────────────────────
    fprintf('Reading %s ...\n', tl_file);
    tl = readtable(tl_file);
    fprintf('  -> %d rows\n', height(tl));

    fprintf('Reading %s ...\n', sign_file);
    sg = readtable(sign_file);
    fprintf('  -> %d rows\n', height(sg));

    % ── Compute time axes (seconds from start) ────────────────────
    tl_t0 = tl.timestamp(1);
    tl_time = tl.timestamp - tl_t0;

    sg_t0 = sg.timestamp(1);
    sg_time = sg.timestamp - sg_t0;

    % ── Split TL latencies by mode ────────────────────────────────
    is_det = strcmp(tl.mode, 'DETECTION');
    is_trk = strcmp(tl.mode, 'TRACKING');

    det_latencies = tl.detection_latency_ms(is_det);
    trk_latencies = tl.tracking_latency_ms(is_trk);
    sign_latencies = sg.latency_ms;

    % ══════════════════════════════════════════════════════════════
    % FIGURE 1 — Latency Bar Chart
    % ══════════════════════════════════════════════════════════════
    fig1 = figure('Name', 'Perception Latency', ...
                  'Position', [100, 300, 700, 500], ...
                  'Color', 'w');

    means = [mean(det_latencies), mean(trk_latencies), mean(sign_latencies)];
    stds  = [std(det_latencies),  std(trk_latencies),  std(sign_latencies)];
    labels = {'TL Detection', 'TL Tracking', 'Sign Detection'};
    colors = [C.tl_det; C.tl_trk; C.sign];

    b = bar(1:3, means, 0.6, 'FaceColor', 'flat', 'EdgeColor', 'none');
    b.CData = colors;
    hold on;

    % Error bars (±1σ)
    er = errorbar(1:3, means, stds, 'LineStyle', 'none', ...
                  'Color', C.err, 'LineWidth', 1.5, 'CapSize', 12);

    % Real-time threshold line (50 ms)
    yline(50, '--', '50 ms real-time', ...
          'Color', [0.6 0.2 0.2], 'LineWidth', 1.2, ...
          'LabelHorizontalAlignment', 'left', ...
          'FontSize', 10, 'FontWeight', 'bold');

    % Stats annotation
    stats_str = '';
    for k = 1:3
        switch k
            case 1, data = det_latencies;
            case 2, data = trk_latencies;
            case 3, data = sign_latencies;
        end
        stats_str = [stats_str, sprintf('%s:\n', labels{k}), ...
                     sprintf('  Mean=%.1f  Std=%.1f  P95=%.1f  Max=%.1f\n', ...
                     mean(data), std(data), prctile(data, 95), max(data))];
    end
    annotation('textbox', [0.58, 0.60, 0.38, 0.30], ...
               'String', stats_str, ...
               'FitBoxToText', 'on', ...
               'BackgroundColor', [1 1 1 0.85], ...
               'EdgeColor', [0.7 0.7 0.7], ...
               'FontSize', 8, ...
               'FontName', 'Consolas', ...
               'Interpreter', 'none');

    set(gca, 'XTickLabel', labels, 'FontSize', 11, ...
             'Box', 'off', 'Color', C.bg);
    ylabel('Latency (ms)', 'FontSize', 12, 'FontWeight', 'bold');
    title('Perception Pipeline Latency', ...
          'FontSize', 14, 'FontWeight', 'bold');
    grid on;
    set(gca, 'GridColor', C.grid, 'GridAlpha', 0.7);
    hold off;

    % ══════════════════════════════════════════════════════════════
    % FIGURE 2 — Detection Rate Over Time
    % ══════════════════════════════════════════════════════════════
    fig2 = figure('Name', 'Detection Rate', ...
                  'Position', [150, 250, 900, 450], ...
                  'Color', 'w');

    plot(tl_time, tl.detection_rate_pct, '-', ...
         'Color', C.tl_det, 'LineWidth', 1.8);
    hold on;
    plot(sg_time, sg.detection_rate_pct, '-', ...
         'Color', C.sign, 'LineWidth', 1.8);

    % Final values annotation
    tl_final = tl.detection_rate_pct(end);
    sg_final = sg.detection_rate_pct(end);

    % Horizontal reference line
    yline(80, ':', 'Target 80%', ...
          'Color', [0.4 0.4 0.4], 'LineWidth', 1.0, ...
          'LabelHorizontalAlignment', 'left', ...
          'FontSize', 9);

    % Final value markers
    plot(tl_time(end), tl_final, 'o', ...
         'MarkerSize', 8, 'MarkerFaceColor', C.tl_det, ...
         'MarkerEdgeColor', 'w', 'LineWidth', 1.5);
    text(tl_time(end), tl_final + 3, sprintf('%.1f%%', tl_final), ...
         'Color', C.tl_det, 'FontSize', 10, 'FontWeight', 'bold', ...
         'HorizontalAlignment', 'right');

    plot(sg_time(end), sg_final, 'o', ...
         'MarkerSize', 8, 'MarkerFaceColor', C.sign, ...
         'MarkerEdgeColor', 'w', 'LineWidth', 1.5);
    text(sg_time(end), sg_final - 3, sprintf('%.1f%%', sg_final), ...
         'Color', C.sign, 'FontSize', 10, 'FontWeight', 'bold', ...
         'HorizontalAlignment', 'right');

    xlabel('Time (s)', 'FontSize', 12, 'FontWeight', 'bold');
    ylabel('Detection Rate (%)', 'FontSize', 12, 'FontWeight', 'bold');
    title('Cumulative Detection Rate', ...
          'FontSize', 14, 'FontWeight', 'bold');
    legend({'Traffic Light', 'Sign Detection'}, ...
           'Location', 'southeast', 'FontSize', 10);
    ylim([0, 105]);
    grid on;
    set(gca, 'GridColor', C.grid, 'GridAlpha', 0.7, ...
             'FontSize', 11, 'Box', 'off', 'Color', C.bg);
    hold off;

    % ══════════════════════════════════════════════════════════════
    % FIGURE 3 — Confidence Over Time
    % ══════════════════════════════════════════════════════════════
    fig3 = figure('Name', 'Detection Confidence', ...
                  'Position', [200, 200, 900, 450], ...
                  'Color', 'w');

    % ── Left y-axis: TL colour ratio ──
    yyaxis left;
    plot(tl_time, tl.best_color_ratio, '-', ...
         'Color', C.tl_det, 'LineWidth', 1.2);
    hold on;
    % Running mean (50-sample window)
    if height(tl) >= 50
        tl_smooth = movmean(tl.best_color_ratio, 50);
        plot(tl_time, tl_smooth, '-', ...
             'Color', C.tl_det * 0.7, 'LineWidth', 2.0);
    end
    ylabel('TL Colour Ratio', 'FontSize', 12, 'FontWeight', 'bold');
    set(gca, 'YColor', C.tl_det);
    ylim([0, 1.05]);

    % ── Right y-axis: Sign vote confidence ──
    yyaxis right;
    plot(sg_time, sg.vote_confidence, '-', ...
         'Color', C.sign, 'LineWidth', 1.2);
    hold on;
    if height(sg) >= 50
        sg_smooth = movmean(sg.vote_confidence, 50);
        plot(sg_time, sg_smooth, '-', ...
             'Color', C.sign * 0.7, 'LineWidth', 2.0);
    end
    ylabel('Sign Vote Confidence', 'FontSize', 12, 'FontWeight', 'bold');
    set(gca, 'YColor', C.sign);
    ylim([0, 1.05]);

    % ── Shared formatting ──
    xlabel('Time (s)', 'FontSize', 12, 'FontWeight', 'bold');
    title('Detection Confidence', ...
          'FontSize', 14, 'FontWeight', 'bold');
    legend({'TL Colour Ratio', 'TL (smoothed)', ...
            'Sign Confidence', 'Sign (smoothed)'}, ...
           'Location', 'northeast', 'FontSize', 9);
    grid on;
    set(gca, 'GridColor', C.grid, 'GridAlpha', 0.7, ...
             'FontSize', 11, 'Box', 'off');
    hold off;

    % ══════════════════════════════════════════════════════════════
    % Export to PNG
    % ══════════════════════════════════════════════════════════════
    out_dir = fullfile(kpi_dir, 'plots');
    if ~exist(out_dir, 'dir')
        mkdir(out_dir);
    end

    exportgraphics(fig1, fullfile(out_dir, 'latency_bar_chart.png'), ...
                   'Resolution', 300);
    exportgraphics(fig2, fullfile(out_dir, 'detection_rate.png'), ...
                   'Resolution', 300);
    exportgraphics(fig3, fullfile(out_dir, 'confidence.png'), ...
                   'Resolution', 300);

    fprintf('\n✓ Plots saved to %s\n', out_dir);
    fprintf('  - latency_bar_chart.png\n');
    fprintf('  - detection_rate.png\n');
    fprintf('  - confidence.png\n');
end
